src/main/java/org/zalando/nakadi/service/converter/VersionZeroConverter.java

```diff
@@ -32,6 +43,71 @@
         return CursorConverter.Version.ZERO;
     }
 
+    public List<NakadiCursor> convertBatched(final List<SubscriptionCursorWithoutToken> cursors) throws
+            InvalidCursorException, InternalNakadiException, NoSuchEventTypeException, ServiceUnavailableException {
+        final NakadiCursor[] result = new NakadiCursor[cursors.size()];
+        for (int idx = 0; idx < cursors.size(); ++idx) {
+            final SubscriptionCursorWithoutToken cursor = cursors.get(idx);
+            if (Cursor.BEFORE_OLDEST_OFFSET.equalsIgnoreCase(cursor.getOffset())) {
+                // Preform begin checks afterwards to optimize calls
+                continue;
+            }
+            if (!NUMBERS_ONLY_PATTERN.matcher(cursor.getOffset()).matches()) {
+                throw new InvalidCursorException(CursorError.INVALID_OFFSET, cursor);
+            }
+            if (cursor.getOffset().startsWith("-")) {
+                result[idx] = new NakadiCursor(
+                        timelineService.getFakeTimeline(eventTypeCache.getEventType(cursor.getEventType())),
+                        cursor.getPartition(),
+                        cursor.getOffset());
+            } else {
+                result[idx] = new NakadiCursor(
+                        timelineService.getFakeTimeline(eventTypeCache.getEventType(cursor.getEventType())),
+                        cursor.getPartition(),
+                        StringUtils.leftPad(cursor.getOffset(), VERSION_ZERO_MIN_OFFSET_LENGTH, '0'));
+            }
+        }
+        // now it is time for massive convert.
+        final LinkedHashMap<SubscriptionCursorWithoutToken, NakadiCursor> beginsToConvert = new LinkedHashMap<>();
+        final Map<SubscriptionCursorWithoutToken, Timeline> cursorTimelines = new HashMap<>();
+        final Map<TopicRepository, List<SubscriptionCursorWithoutToken>> repos = new HashMap<>();
+        for (int i = 0; i < result.length; ++i) {
+            if (null == result[i]) { // cursor requires database hit
+                final SubscriptionCursorWithoutToken cursor = cursors.get(i);
+                final Timeline timeline = timelineService.getActiveTimelinesOrdered(cursor.getEventType()).get(0);
+                final TopicRepository topicRepo = timelineService.getTopicRepository(timeline);
+                beginsToConvert.put(cursor, null);
```

Comment:
In order to do 2 things: 
 - collect value that is to be converted
 - preserve order of cursors to be converted